solutions/USACO feb19-cowland.mdx

```diff
@@ -154,4 +154,270 @@ int main()
+                    continue;
+                }
+                visited.add(curr[0]);
+                height[curr[0]] = curr[2];
+                firstOcc[curr[0]] = eulerTour.size() - 1;  // -1 because of that .add(curr[0]) above
+                if (curr[0] != 0) {
+                    frontier.addFirst(new int[]{curr[1], parent[curr[1]], curr[2] - 1});
+                }
+                boolean isLeaf = true;  // i know i could like check the amt of neighbors but this is somehow faster
+                for (int n : neighbors[curr[0]]) {
+                    if (n != curr[1]) {  // visit all children (don't process the parent)
+                        parent[n] = curr[0];
+                        frontier.addFirst(new int[]{n, curr[0], curr[2] + 1});
+                        isLeaf = false;
+                    }
+                }
+                if (isLeaf) {
+                    eulerTour.add(curr[0]);
+                }
+            }
+        }
+    }
+
+    public static void main(String[] args) throws IOException {
+        long start = System.currentTimeMillis();
+        BufferedReader read = new BufferedReader(new FileReader("cowland.in"));
+        StringTokenizer initial = new StringTokenizer(read.readLine());
+        int placeNum = Integer.parseInt(initial.nextToken());
+        int queryNum = Integer.parseInt(initial.nextToken());
+        int[] initialVals = Arrays.stream(read.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
+        ArrayList<Integer>[] neighbors = new ArrayList[placeNum];
+        for (int p = 0; p < placeNum; p++) {
+            neighbors[p] = new ArrayList<>();
+        }
+        for (int i = 0; i < placeNum - 1; i++) {
+            int[] route = Arrays.stream(read.readLine().split(" ")).mapToInt(j -> Integer.parseInt(j) - 1).toArray();
+            neighbors[route[0]].add(route[1]);
+            neighbors[route[1]].add(route[0]);
+        }
+        AmusementPark park = new AmusementPark(neighbors, initialVals);
+        PrintWriter written = new PrintWriter("cowland.out");
+        for (int q = 0; q < queryNum; q++) {
+            int[] query = Arrays.stream(read.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
+            switch (query[0]) {
+                case 1:
+                    park.set(query[1] - 1, query[2]);
+                    break;
+                case 2:
+                    int happiness = park.query(query[1] - 1, query[2] - 1);
+                    // System.out.println(happiness);  // <- adding this makes the thing too slow lol
+                    written.println(happiness);
+                    break;
+                default:
+                    throw new IllegalArgumentException("what kind of query is " + Arrays.toString(query) + " lol");
+            }
+        }
+        written.close();
+        System.out.printf("WOOOOOASDPOFIJASDFIOJASPODI it took %d ms%n", System.currentTimeMillis() - start);
```

Comment:
o_O

Reply:
Oh yeah, sorry about that. I usually have a timer in my solutions and I add some flavor text.